haystack/tools/serde_utils.py

```diff
@@ -2,28 +2,37 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
-from typing import Any, Union
+from typing import TYPE_CHECKING, Any, Optional, Union
 
 from haystack.core.errors import DeserializationError
 from haystack.core.serialization import import_class_by_name
 from haystack.tools.tool import Tool
 from haystack.tools.toolset import Toolset
 
+if TYPE_CHECKING:
+    from haystack.tools import ToolsType
 
-def serialize_tools_or_toolset(
-    tools: Union[Toolset, list[Tool], None],
-) -> Union[dict[str, Any], list[dict[str, Any]], None]:
+
+def serialize_tools_or_toolset(tools: "Optional[ToolsType]") -> Union[dict[str, Any], list[dict[str, Any]], None]:
     """
-    Serialize a Toolset or a list of Tools to a dictionary or a list of tool dictionaries.
+    Serialize tools or toolsets to dictionaries.
 
-    :param tools: A Toolset, a list of Tools, or None
-    :returns: A dictionary, a list of tool dictionaries, or None if tools is None
+    :param tools: A Toolset, a list of Tools and/or Toolsets, or None
+    :returns: Serialized representation preserving Tool/Toolset boundaries when provided
     """
     if tools is None:
         return None
     if isinstance(tools, Toolset):
         return tools.to_dict()
-    return [tool.to_dict() for tool in tools]
+    if isinstance(tools, list):
+        serialized: list[dict[str, Any]] = []
+        for item in tools:
+            if isinstance(item, (Toolset, Tool)):
+                serialized.append(item.to_dict())
+            else:
+                raise TypeError("Items in the tools list must be Tool or Toolset instances.")
+        return serialized
+    raise TypeError("tools must be Toolset, list[Union[Tool, Toolset]], or None")
 
 
 def deserialize_tools_or_toolset_inplace(data: dict[str, Any], key: str = "tools") -> None:
```

Comment:
We should update the docstrings of this function to reflect the new type serialized tools can have. E.g. `Deserialize a list of Tools or a Toolset in a dictionary inplace.` is no longer fully correct